### PR TITLE
Npm packages

### PR DESCRIPTION
**Pull request by RobinNagpal**

### CODE REVIEW

File: src/aave-npm-packages.yaml [hunk 1]
```diff
@@ -7,35 +7,415 @@ created: "2022-08-02T04:15:13.911Z"
 categories:
   - developer
   - engineering
-thumbnail: https://d31h13bdjwgzxs.cloudfront.net/guides/eth/aave/AAVE_NPM_Packages.png
+thumbnail: https://d31h13bdjwgzxs.cloudfront.net/guides/eth/aave/Borrowing_in_aave.png
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]: No need to change the image

File: src/aave-npm-packages.yaml [hunk 2]
```diff
@@ -7,35 +7,415 @@ created: "2022-08-02T04:15:13.911Z"
...
+
+     You can install @aave/aave-v3-core as an npm package to your hardhat or truffle workspace to get access to contracts and 
+     interfaces provided by Aave V3
+
+     The repository uses Docker Compose to manage sensitive keys and load the configuration. 
+     Before any action like test or deploy, you must run docker-compose up to start the contracts-env container, and then connect to the container console through docker-compose exec contracts-env bash.
+     You can run a full test suite using Docker Compose commands.
+
+     ## Aave coreV2
+     This is the previous version of Aave core V3. and so it has features similar to V3.
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]:
```suggestion
     package contains smart contracts for V2.
```

File: src/aave-npm-packages.yaml [hunk 3]
```diff
@@ -7,35 +7,415 @@ created: "2022-08-02T04:15:13.911Z"
...
+     ## Aave coreV3
+     This package contains the smart contracts source code and markets configuration related to Aave Protocol V3. 
+     It uses Docker Compose and Hardhat as development environment for compilation, testing and deployment tasks.
+
+     You can install @aave/aave-v3-core as an npm package to your hardhat or truffle workspace to get access to contracts and 
+     interfaces provided by Aave V3
+
+     The repository uses Docker Compose to manage sensitive keys and load the configuration. 
+     Before any action like test or deploy, you must run docker-compose up to start the contracts-env container, and then connect to the container console through docker-compose exec contracts-env bash.
+     You can run a full test suite using Docker Compose commands.
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]:
````suggestion

You can install `@aave/core-v3` as an NPM package in your Hardhat or Truffle project to import the contracts and interfaces:

`npm install @aave/core-v3`

Import at Solidity files:

```
import {IPool} from "@aave/core-v3/contracts/interfaces/IPool.sol";

contract Misc {

  function supply(address pool, address token, address user, uint256 amount) public {
    IPool(pool).supply(token, amount, user, 0);
    {...}
  }
}
```

The JSON artifacts with the ABI and Bytecode are also included into the bundled NPM package at `artifacts/` directory.

Import JSON file via Node JS `require`:

```
const PoolV3Artifact = require('@aave/core-v3/artifacts/contracts/protocol/pool/Pool.sol/Pool.json');

// Log the ABI into console
console.log(PoolV3Artifact.abi)
```
````

File: src/aave-npm-packages.yaml [hunk 4]
```diff
@@ -7,35 +7,415 @@ created: "2022-08-02T04:15:13.911Z"
...
 steps:
-  - uuid: 18f145d1-e807-4da3-98f3-57d010354ce0
-    name: List of packages
+  - uuid: 5db956ef-4e00-4ac1-8c25-64c5dbebfa73
+    name: Aave core and stake
     content: |
-      List of packages
-      See here - https://www.npmjs.com/search?q=%40aave
+
+     ## Aave coreV3
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]:
```suggestion
     ## [@aave/core-v3](https://www.npmjs.com/package/@aave/core-v3)
```

File: src/aave-npm-packages.yaml [hunk 5]
```diff
@@ -7,35 +7,415 @@ created: "2022-08-02T04:15:13.911Z"
...
+     It uses Docker Compose and Hardhat as development environment for compilation, testing and deployment tasks.
+
+     You can install @aave/aave-v3-core as an npm package to your hardhat or truffle workspace to get access to contracts and 
+     interfaces provided by Aave V3
+
+     The repository uses Docker Compose to manage sensitive keys and load the configuration. 
+     Before any action like test or deploy, you must run docker-compose up to start the contracts-env container, and then connect to the container console through docker-compose exec contracts-env bash.
+     You can run a full test suite using Docker Compose commands.
+
+     ## Aave coreV2
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]:
```suggestion
     ## [@aave/protocol-v2](https://www.npmjs.com/package/@aave/protocol-v2)
```

File: src/aave-npm-packages.yaml [hunk 6]
```diff
@@ -7,35 +7,415 @@ created: "2022-08-02T04:15:13.911Z"
...
+     This is the previous version of Aave core V3. and so it has features similar to V3.
+     The repository uses Docker Compose to manage sensitive keys and load the configuration. 
+     Prior any action like test or deploy, you must run docker-compose up to start the contracts-env container, and then connect to the container console via docker-compose exec contracts-env bash.
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]:
```suggestion
     Just like @aave/core-v3, you can  install @aave/protocol-v2 as an NPM package by running `npm install @aave/protocol-v2`
     
     You can import Solidity and ABI files, just like the V3 files shown above. 
```

File: src/aave-npm-packages.yaml [hunk 7]
```diff
@@ -7,35 +7,415 @@ created: "2022-08-02T04:15:13.911Z"
...
+
+     ## Aave Stake
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]: Use the correct npm package name and add couple of paragraphs for this.

https://www.npmjs.com/package/@aave/aave-stake

In the first paragraph explain about Safety module and in the second, high level about this repository

File: src/aave-npm-packages.yaml [hunk 8]
```diff
@@ -7,35 +7,415 @@ created: "2022-08-02T04:15:13.911Z"
...
+          - content: None of these
+            key: D
+ 
+
+  - uuid: 16959203-96fa-4b9a-ac7b-67e0905d7533
+    name: Aave Lens Protocol
     content: |
-      https://www.npmjs.com/package/@aave/periphery-v3
+
+     ## Aave Lens Protocol
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]: We can skip lens protocol and move https://www.npmjs.com/package/@aave/aave-stake to this step

File: src/aave-npm-packages.yaml [hunk 9]
```diff
@@ -11,30 +11,363 @@ publishStatus: Live
...
+
+
+  - uuid: 48b327fc-9ad8-481c-8948-989dad8020e1
+    name: |
+     @aave/protocol-js, @aave/safety-module, @aave/protocol-v2
     content: |
-      https://www.npmjs.com/package/@aave/math-utils
+     
+     ## @aave/protocol-js
+     AAVE is a decentralized non-custodial liquidity market protocol where users can participate as depositors or borrowers.
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]: Do you feel this line `AAVE is a decentralized non-custodial liquidity market protocol where users can participate as depositors or borrowers. ` add value to the user here? 

is this related to `@aave/protocol-js`? Why specifically here?

File: src/aave-npm-packages.yaml [hunk 10]
```diff
@@ -11,30 +11,363 @@ publishStatus: Live
...
+     @aave/protocol-js, @aave/safety-module, @aave/protocol-v2
     content: |
-      https://www.npmjs.com/package/@aave/math-utils
+     
+     ## @aave/protocol-js
+     AAVE is a decentralized non-custodial liquidity market protocol where users can participate as depositors or borrowers. 
+     The AAVE Protocol is a set of open source smart contracts which facilitate the lending and borrowing of user funds. These contracts, 
+     and all user transactions/balances are stored on a public ledger called a blockchain, making them accessible to anyone.
+
+     ethers js, something we discussed about earlier, is a peer dependancy (PeerDependency, it is not automatically installed. Instead, the code that includes the package must include it as its dependency.) for protocol-js.
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]: How does this line connect to the previous line?

`ethers js, something we discussed about earlier, is a peer dependancy`

File: src/aave-npm-packages.yaml [hunk 11]
```diff
@@ -11,30 +11,363 @@ publishStatus: Live
...
+
+     The aave-js data formatting methods are a layer beyond graphql which wraps protocol data into more usable formats. Each method will require inputs from AAVE subgraph queries.
+     As you can see, we have already discussed about formatting packages (@aave/math-utils), these features are enclosed in this package. It has data formatting services which allow the querying of 
+     User data and reserve data.
+     It also encloses various transaction methods and the **Lending pool V2**, of which we have talked about in detail. For a short summary,
+     Lending pool is just a smart contract which acts as a reserve (pool) of assets, lent by liquidity providers for the protocol to use
+     for loaning. It contains methods like deposit, borrow, repay, withdraw, etc which can be called by users via smartcontracts.
+     It also encloses other vast concepts like aave governance.
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]: We wrote two paragraphs under `@aave/protocol-js` but nothing actually is related to `@aave/protocol-js` ?

File: src/aave-npm-packages.yaml [hunk 12]
```diff
@@ -11,30 +11,363 @@ publishStatus: Live
...
+     The module will also cover BAL tokens and trading fees. Plus, it will reduce the impact of a Shortfall Event on the AAVE token.
+     In the event of a Shortfall, a portion of the locked AAVE is auctioned off to be sold in order to make up for the deficit. The SM has a built-in backstop mechanism to prevent too 
+     much AAVE from entering the open market, which would lower the value of AAVE. By locking AAVE into the SM, participants are accepting the possibility of a Shortfall Event occurring, 
+     in exchange for rewards in the form of Safety Incentives (SI).
+    
+     AAVE holders can deposit their tokens into the Staking Mechanism to contribute to the safety of the protocol and receive incentives. In return, they will receive a tokenized position
+     that can be freely moved within the underlying network. The holder of the tokenized position can redeem their share from the SM at any time, triggering a cooldown period of 
+     one week (which can be further extended by the governance, mentioned in the protocol-js).
+
+     ## @aave/protocol-v2
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]: How is this different from @aave/coreV2 discussed in step1 ?

File: src/aave-npm-packages.yaml [hunk 13]
```diff
@@ -11,30 +11,363 @@ publishStatus: Live
...
+
     stepItems: []
 
+  - uuid: 0ee3b981-2e16-46dd-9400-fbc3f3a58746
+    name: Evaluation
+    content: |
+    stepItems:
+      - uuid: 881f3900-a654-4868-8d62-b80153494bf8
+        content: | 
+         @aave/protocol-js encloses which of these?
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]: What does `encloses` mean?

File: src/aave-npm-packages.yaml [hunk 14]
```diff
@@ -11,30 +11,363 @@ publishStatus: Live
...
+        answerKeys:
+          - A
+          - B
+          - C
+        type: MultipleChoice
+        choices:
+          - content: aave governance
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]:
```suggestion
          - content: Aave governance
```

File: src/aave-npm-packages.yaml [hunk 15]
```diff
@@ -11,30 +11,363 @@ publishStatus: Live
...
+            key: A
+          - content: aave lending pool
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]:
```suggestion
          - content: Aave lending pool
```
correct casing for the other options also

File: src/aave-npm-packages.yaml [hunk 16]
```diff
@@ -11,30 +11,363 @@ publishStatus: Live
...
+     The @aave/math-utils data formatting methods act as a layer on top of the chain data. 
+     The use-cases range from "human readable formatting" to "approximating accrual over time."
+     We know users interact with the aave protocol using smartcontracts, @aave/math-utils package is a collection of methods to take raw input data 
+     from these contracts, and format to use on a frontend interface such as Aave Ui or Aave info.
+
+     It formats two types of data, raw and indexed contract data, for front-end usage.
+      * ethers.js: 
+        ethers.js is considered raw contract data. ethers.js is a library for interacting with Ethereum and other EVM compatible blockchains.
+
+      * subgraph:
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]: How is this `subgraph` point related to @aave/math-utils?

File: src/aave-npm-packages.yaml [hunk 17]
```diff
@@ -11,30 +11,363 @@ publishStatus: Live
...
+        It is an indexed data type. Subgraphs work by indexing events that come from smart contracts. 
+        This allows for data to be queried via a graphql endpoint. In other words, for each network where the protocol is deployed, there is a subgraph that corresponds to it.
+      
+      * Caching servers:
+        The Aave UI uses decentralized hosting with IPFS, which means that it doesn't require any API keys to run. However, this also means that only public rpc endpoints can be used, which are quickly rate-limited. 
+        To respond for this, Aave has published a caching server that performs contract queries on one server and then serves the data to all frontend users through a graphQL endpoint.
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml]: How is `Caching servers` related to `@aave/math-utils` ?

File: src/aave-npm-packages.yaml (hunk 9 above)
Comment (mviswanathsai) [reply src/aave-npm-packages.yaml]: My thought process was that since protocol-js encompasses the Aave protocol as a whole, it might be of value to add what the set of smart contracts provide the users with.

File: src/aave-npm-packages.yaml (hunk 10 above)
Comment (mviswanathsai) [reply src/aave-npm-packages.yaml]: discussed earlier in math-utils

     It formats two types of data, raw and indexed contract data, for front-end usage.
      * ethers.js: 
        ethers.js is considered raw contract data. ethers.js is a library for interacting with Ethereum and other EVM compatible blockchains.

File: src/aave-npm-packages.yaml (hunk 16 above)
Comment (mviswanathsai) [reply src/aave-npm-packages.yaml]: subgraphs and caching servers are used to collect input data for the math utils package to format

File: src/aave-npm-packages.yaml [hunk 18]
```diff
@@ -11,30 +11,591 @@ publishStatus: Live
...
+        choices:
+           - content: create
+             key: A
+           - content: queue
+             key: B
+           - content: submitVote
+             key: C
+           - content: delegate
+             key: D
+      - uuid: 9a8229eb-f548-4596-b938-7d7f63dc6adb
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml:194]: Is 'C' the right option?  setUsageAsCollateral see to set something rather then determine

Comment (RobinNagpal) [reply src/aave-npm-packages.yaml:194]: Please reword the options for better clarity

File: src/aave-npm-packages.yaml [hunk 19]
```diff
@@ -11,30 +11,591 @@ publishStatus: Live
...
+          - A
+          - B
+          - D
+        type: MultipleChoice
+        choices:
+          - content: Swap the collateral asset for another asset
+            key: A
+          - content: Use the collateral to repay loan
+            key: B
+          - content: Withdraw all of the collateral on command
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml:235]: Reword this option `Withdraw all of the collateral on command` as its not clear

File: src/aave-npm-packages.yaml [hunk 20]
```diff
@@ -11,30 +11,591 @@ publishStatus: Live
...
+            key: A
+          - content: Address of the user who deposits/ borrows/ etc.,.
+            key: B
+          - content: Could be both A and B
+            key: C
+          - content: Neither A nor B
+            key: D
+
+      - uuid: 4aa31aea-1718-433c-9ebf-44cf22b442c7
+        content: Methods that concern with collaterals, allow the user to do which of the following?
```
Comment (RobinNagpal) [inline src/aave-npm-packages.yaml:224]:
```suggestion
        content: What operations related to collaterals can the user perform using `@aave/protocol-js`?
```
